activation/handler.go

```diff
@@ -172,27 +172,58 @@ func (h *Handler) syntacticallyValidate(ctx context.Context, atx *wire.Activatio
 	return nil
 }
 
+func (h *Handler) commitment(ctx context.Context, atx *wire.ActivationTxV1) (types.ATXID, error) {
+	if atx.PrevATXID == types.EmptyATXID {
+		return *atx.CommitmentATXID, nil
+	}
+	return atxs.CommitmentATX(h.cdb, atx.SmesherID)
+}
+
+func (h *Handler) previous(ctx context.Context, id types.ATXID) (*types.ActivationTx, error) {
+	var blob sql.Blob
+	if err := atxs.LoadBlob(ctx, h.cdb, id[:], &blob); err != nil {
+		return nil, err
+	}
+
+	if blob.Bytes == nil {
+		// if the blob is empty, it means the ATX is golden - fetch it from the DB
+		vatx, err := atxs.Get(h.cdb, id)
+		if err != nil {
+			return nil, fmt.Errorf("fetching golden previous atx: %w", err)
+		}
+		return vatx.ActivationTx, nil
+	}
+
+	var atx wire.ActivationTxV1
+	if err := codec.Decode(blob.Bytes, &atx); err != nil {
+		return nil, fmt.Errorf("decoding previous atx: %w", err)
+	}
+	return wire.ActivationTxFromWireV1(&atx, blob.Bytes...), nil
+}
+
 func (h *Handler) syntacticallyValidateDeps(
 	ctx context.Context,
 	atx *wire.ActivationTxV1,
 ) (leaves uint64, effectiveNumUnits uint32, proof *mwire.MalfeasanceProof, err error) {
-	var commitmentATX types.ATXID
+	commitmentATX, err := h.commitment(ctx, atx)
+	if err != nil {
+		return 0, 0, nil, fmt.Errorf("commitment atx for %s not found: %w", atx.SmesherID, err)
+	}
 
 	if atx.PrevATXID == types.EmptyATXID {
 		if err := h.nipostValidator.InitialNIPostChallengeV1(&atx.NIPostChallengeV1, h.cdb, h.goldenATXID); err != nil {
 			return 0, 0, nil, err
 		}
 		effectiveNumUnits = atx.NumUnits
-		commitmentATX = *atx.CommitmentATXID
 	} else {
-		commitmentATX, err = atxs.CommitmentATX(h.cdb, atx.SmesherID)
+		previous, err := h.previous(ctx, atx.PrevATXID)
```

Comment:
I almost commented, "why not use atxs.Get?" 😅

Can you add a clarifying comment that we need to decode the previous ATX to get the post NumUnits value so we can calculate effective numUnits?

Reply:
I explained what's happening in a comment